File: Algorithms/Two Pointer/3. Longest Substring Without Repeating Characters.py
class Solution(object):
	def lengthOfLongestSubstring(self, s):
		length = len(s)
		charset = {}
		maxLength = 0
		start = 0
		for right in range(len(s)):
			char = s[right]
			if(char in charset and charset[char] >= start):
				start = charset[char] + 1	
			else:
				maxLength = max(maxLength, right - start + 1)
			charset[char] = right
		return maxLength

File: Algorithms/Two Pointer/3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
	vector<int> hash(128, 0);
	int left = 0, right = 0, count = 0, window = 0;
	while(right < s.size())
	{
		if(hash[s[right++]]++ == 1) count += 1;
		while(count > 0)
		{
			if(hash[s[left++]]-- == 2) count -= 1;	
		}
		window = max(window, right - left);
	}
	return window;
    }
};
